discovery/azure/azure.go

```diff
@@ -192,55 +212,70 @@ func createAzureClient(cfg SDConfig) (azureClient, error) {
 	resourceManagerEndpoint := env.ResourceManagerEndpoint
 
 	var c azureClient
+	var credential azcore.TokenCredential
 
-	var spt *adal.ServicePrincipalToken
+	var userAgentPolicy = NewUserAgentPolicy(userAgent)
 
 	switch cfg.AuthenticationMethod {
 	case authMethodManagedIdentity:
-		spt, err = adal.NewServicePrincipalTokenFromManagedIdentity(resourceManagerEndpoint, &adal.ManagedIdentityOptions{ClientID: cfg.ClientID})
+		managedIdentityCredential, err := azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{ClientOptions: policy.ClientOptions{
+			Cloud: cloud.Configuration{
+				ActiveDirectoryAuthorityHost: resourceManagerEndpoint, Services: map[cloud.ServiceName]cloud.ServiceConfiguration{},
+			},
+			PerCallPolicies: []policy.Policy{userAgentPolicy},
+		}, ID: azidentity.ClientID(cfg.ClientID)})
 		if err != nil {
 			return azureClient{}, err
 		}
+		credential = azcore.TokenCredential(managedIdentityCredential)
 	case authMethodOAuth:
-		oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, cfg.TenantID)
-		if err != nil {
-			return azureClient{}, err
-		}
-
-		spt, err = adal.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, string(cfg.ClientSecret), resourceManagerEndpoint)
+		secretCredential, err := azidentity.NewClientSecretCredential(cfg.TenantID, cfg.ClientID, string(cfg.ClientSecret), &azidentity.ClientSecretCredentialOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloud.Configuration{
+					ActiveDirectoryAuthorityHost: activeDirectoryEndpoint, Services: map[cloud.ServiceName]cloud.ServiceConfiguration{},
+				},
+				PerCallPolicies: []policy.Policy{userAgentPolicy},
+			},
+		})
 		if err != nil {
 			return azureClient{}, err
 		}
+		credential = azcore.TokenCredential(secretCredential)
 	}
 
-	client, err := config_util.NewClientFromConfig(cfg.HTTPClientConfig, "azure_sd")
+	httpClient, err := config_util.NewClientFromConfig(cfg.HTTPClientConfig, "azure_sd")
 	if err != nil {
 		return azureClient{}, err
 	}
-	sender := autorest.DecorateSender(client)
-	preparer := autorest.WithUserAgent(userAgent)
```

Comment:
As it is not possible do customization of the User-Agent directly, I had to create dedicated Policy

Reply:
Since you only need to add version info of prometheus, you could use `ApplicationID` (in `Telemetry` of `ClientOptions`) to config your application's specific `User-Agent` field. Please refer: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.3.0/policy#ClientOptions.

Reply:
Thank you !